### PR TITLE
ch-image: add git build cache

### PR DESCRIPTION
**Pull request by j-ogas**
Addresses #583.

### CODE REVIEW

**Review by reidpr — COMMENTED** [review]
Fantastic work! I refactored quite a bit. Remaining to-do, likely incomplete, brain dump from notes I took over the past couple weeks:
* [x] Fix ARG hits not setting variable
* [x] Figure out actual Git, git2dot, dot minimum versions needed.
* [x] Pedantic-fail cache tests if Git not installed.
* [x] Add environmental variable for build cache mode, for testing.
* [ ] Currently pull to specific directory (not storage) is gone. Shall we restore it? Does anyone use it? If we don't restore it, we could re-use the argument for the destination ref, to work like `push`.
* [x] Ready marking. I'm leaning towards branch names, e.g. `alpine+3.9+Z` (not ready) and `alpine+3.9` (ready)
* [ ] Exclude metadata file (`git.pickle`) and `.git` from export and push.
* [x] new issue: Save/restore ACLs and xattrs. I didn't see anything in the standard library to do this. Maybe we can use `ctypes`?
* [x] new issue: Put Git repo out-of-band in a separate directory. With in-band as currently, the root of the image can't be a Git repo.
* [x] new issue: Deal with hard links. Will also need to fix their size being double-counted.
* [ ] Upgrade v2 storage to v3. Existing image directories will need to be renamed.
* [x] new issue: Add `--debug` option to provide stack trace for fatal errors (this is already implemented in the PR).
* [ ] `--dry-run` is gone. It only sort of worked anyway, because often instructions depended on earlier instructions actually happening. Add a new issue and update the docs that it's gone?
* [ ] CI: add cache enabled/disabled to matrix.
* [ ] Decide whether to make enabled the default upon release. It's quite a big change and maybe it should be available but not the default for one release. Performance tests may illuminate this decision.
* [ ] Document that image root can't contain any Git metadata, and verify this.
* [ ] Implement cache for all instructions and test. `COPY` might need to build up the file list in `prepare()`.
* [x] new issue: Clean up messy globals in `build.py`.
* [ ] fakeroot might be getting initialized unnecessarily on every checkout; looks into this.
* [ ] Multi-stage: probably need to append the `stage_n` to branch name too?
* [ ] Remove `HEAD ->` from `--tree` output?
* [ ] With and without `--force` should not hit each other. I think presently they do.
* [ ] `pull` does not check the cache. I think what's needed is to split the download of the manifest, which will give us the state ID, from the rest.
* [ ] Is there a default `.gitignore` we need to override?
* [ ] Don't fail tests if no `git2dot` or `dot` is present. We could warn instead of failing out, though this would still put text in the output we verify. Could also add another option to silently skip if it's not present.
* [ ] Checkout should do nothing if the right commit is already checked out. I think right now it will delete and re-create the directory.
* [ ] If a worktree already exists with the right name, it should be re-used instead of deleting and re-creating.
* Tests to add:
  * [x] `--gc`; should not crash, check that it removes untagged images.
  * [ ] Various difficult files (see `git_prepare()`).
  * [ ] round-trip a biggish distro through Git and make sure it's all the same.
  * [ ] All hits but new image name.
  * [ ] Multi-stage build.
  * [ ] All hits but directory in `img` doesn't exist.
  * [ ] Restore in quick and full modes?

**Review by j-ogas — COMMENTED** [review]
In addition to refactor/tidying, the following things still need to be done. [ongoing edits to this list by @reidpr]
- [x] Exclude metadata file (`git.pickle`) and `.git*` from export and push.
- [x] Pull now uses build-cache but doesn't have the same branch readiness mechanisms
- [x] Upgrade v2 storage to v3. Existing image directories will need to be renamed
- [x] Document that image root can't contain any Git metadata, and verify this.
- [x] Implement cache for COPY; might need to build up the file list in `prepare()`.
- [x] Is there a default .gitignore we need to override?
- [x] If a worktree already exists with the right name, it should be re-used instead of deleting and re-creating.
- Tests to add:
   - [ ] gc removes unstagged images
   - [ ] Various difficult files (see git_prepare()).
   - [ ] override of ignore patterns (command line vs. our `.gitignore`)
   - [ ] round-trip a biggish distro through Git and make sure it's all the same.
   - [ ] Multi-stage build
   - [ ] All hits but directory in img doesn't exist.
   - [ ] Restore in quick and full modes?
   - [ ] Git metadata at image root (`.git`)

**Comment by reidpr** [comment]
@j-ogas, back at you.

I propose we move the missing tests above to a new PR.

**Review by j-ogas — COMMENTED** [review]
First review pass. Will look at the juicier files this week.

File: lib/build.py [hunk 1]
```diff
@@ -181,137 +157,307 @@ def build_arg_get(arg):
...
       else:
          ch.INFO("--force: init OK & modified %d RUN instructions"
                  % fakeroot_config.inject_ct)
    ch.INFO("grown in %d instructions: %s"
-           % (ml.instruction_ct, images[image_i]))
+           % (ml.instruction_total_ct, ml.inst_prev.image))
+   # FIXME: remove when we're done encouraging people to use the build cache.
+   if (isinstance(bu.cache, bu.Disabled_Cache)):
+      ch.INFO("build slow? consider enabling the new build cache",
+              "https://hpc.github.io/charliecloud/command-usage.html#build-cache")
```
Comment (j-ogas) [inline lib/build.py:185]: The command reference for the build-cache subcommand isn't very useful. It doesn't mention how to enable it nor the dependencies required. Perhaps a link to ch-image, a link to an added faq section, or a different message.

Comment (reidpr) [reply lib/build.py:185]: It should go to the “Build cache” section; if not that’s an error.

File: doc/ch-image_desc.rst [hunk 2]
```diff
@@ -173,6 +189,115 @@ directory with :code:`ch-image reset`.
...
+For example, suppose we have this Dockerfile::
+
+  $ cat a.df
+  FROM alpine:3.9
+  RUN echo foo
+  RUN echo bar
+
+On our first build, we get::
+
+  $ ch-image --bucache=enabled build -t foo -f a.df .
```
[the remark this answers is not in the document]
Comment (reidpr) [reply doc/ch-image_desc.rst:230]: It's just demonstrative. I agree we should not imply that.

File: doc/ch-image_desc.rst [hunk 3]
```diff
@@ -173,6 +189,115 @@ directory with :code:`ch-image reset`.
...
+    1. FROM alpine:3.9
+  [ ... pull chatter omitted ... ]
+    2. RUN echo foo
+  copying image ...
```
[the remark this answers is not in the document]
Comment (reidpr) [reply doc/ch-image_desc.rst:234]: Yes, it’s saying “here’s the instruction I’m going to run”, and the first step is to make somewhere to run it.

**Pull request merged**